### DISCUSSION TITLE
Visual parse tree

### DISCUSSION THREAD

**Discussion by oovm** [pest-parser/pest#940] [Show and tell]
I find the visualization tools in ANTLRWorks very useful:

![image](https://github.com/pest-parser/pest/assets/17541209/798e984e-7d3f-420a-98fe-5eb65548a217)

So I also wrote [pest-viewer](https://github.com/oovm/pest-viewer) for pest:

It can render the following input into a tree diagram based on the pest grammar.

- input

```json
{
    "integer": 1234567890,
    "real": 0.123456789e-12,
    "controls": "\b\f\n\r\t",
    "alpha": "abcdefghijklmnopqrstuvwyz",
    "ALPHA": "ABCDEFGHIJKLMNOPQRSTUVWYZ",
    "array": ["\"", "\\"]
}
```

- output

![json.svg](https://github.com/oovm/pest-viewer/raw/master/projects/pest-viewer/tests/json.svg)

---

This tool is written in pure rust and can theoretically be compiled into wasm.

I'm not sure who is maintaining the web page now. I think it would be helpful for debugging to put this on the home page.

**Comment by tomtau** [comment]
Awesome work! 👏 Yes, it could be helpful to incorporate it in or link it on the main website. It's been primarily @nathanielknight @Jamalam360 @huacnlee @LeoDog896 who worked on the website recently.

**Comment by tomtau** [comment]
@oovm feel free to open a PR to add it to this list https://github.com/pest-parser/awesome-pest

**Comment by Jamalam360** [comment]
It should definitely be possible to get it working on the website if it outputs SVG and can compile to wasm!
